File: src/runtime/php/tokenizer.php
<?php

/*
 * Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

require_once 'Bovinus/token.php';

class Bovinus_Tokenizer {

    public function __construct() {

        $this->_literal = null;
        $this->_literalDelims = array('"', "'");
        $this->_literalEscChar = '\\';
        $this->_separators = array();
    }

    public function setLiteral($literal) {

        $this->_literal = $literal;
        $this->_literalDelims = Bovinus_Literal::$DELIMITERS;
        $this->_literalEscChar = Bovinus_Literal::$ESCAPE_CHAR;
    }

    public function addSeparator($separator) {

        array_push($this->_separators, $separator);
    }

    public function splitAtSeparators($text) {

        $worklist = array(array("text" => $text, "sep" => null));

        foreach ($this->_separators as $sep) {

            $tmp = array();

            foreach ($worklist as $item) {

                if ($item['sep'] != null) {
                    array_push($tmp, $item);
                } else {
                    $tmp = array_merge($tmp, $this->_splitAtSep($item["text"], $sep));
                }
            }

            $worklist = $tmp;
        }

        return $worklist;
    }

    private function _splitAtSep($text, Bovinus_Separator $sep) {

        $res = array();

        $worklist = array(array("text" => $text, "sep" => null, "checked" => FALSE));
        $regex = $sep->getRegex();

        while (TRUE) {

            $tmp = array();
            $doNextStep = FALSE;

            foreach ($worklist as $item) {

                if ($item["checked"]) {
                    array_push($tmp, $item);
                    continue;
                }

                preg_match($regex, $item["text"], $matches);
                $numGroups = count($matches) - 1;

                switch ($numGroups) {

                    case 0:
                    case 1:

                        $doNextStep = TRUE;

                        $newItem = array("text" => $matches[$numGroups], "sep" => $sep, "checked" => TRUE);
                        array_push($tmp, $newItem);
                        break;

                    case 3:

                        $doNextStep = TRUE;

                        $m1 = $matches[1];
                        if (strlen($m1) > 0) {
                            $newItem = array("text" => $m1, "sep" => null, "checked" => FALSE);
                            array_push($tmp, $newItem);
                        }

                        $newItem = array("text" => $matches[2], "sep" => $sep, "checked" => TRUE);
                        array_push($tmp, $newItem);

                        $m3 = $matches[3];
                        if (strlen($m3) > 0) {
                            $newItem = array("text" => $m3, "sep" => null, "checked" => FALSE);
                            array_push($tmp, $newItem);
                        }

                        break;

                    default:

                        $item["checked"] = TRUE;
                        array_push($tmp, $item);
                        break;
                }
            }

            if (!$doNextStep) {
                foreach ($tmp as $el) {
                    array_push($res, array("text" => $el["text"], "sep" => $el["sep"]));
                }
                $res = $this->_unsplitIfInLiteral($res);
                break;
            } else {
                $worklist = $tmp;
            }
        }

        return $res;
    }

    private function _unsplitIfInLiteral($splitResult) {

        if ($this->_literal == null) {
            return $splitResult;
        }

        $res = array();
        $tmp = array();
        $len = count($splitResult);

        for ($i = 0; $i < $len; $i++) {

            $item = $splitResult[$i];

            if ($item["sep"] == null) {
                array_push($tmp, $item);
                continue;
            }

            if ($i > 0 && $i < $len - 1) {

                $left = "";
                for ($j = 0; $j < $i; $j++) {
                    $left .= $splitResult[$j]["text"];
                }

                $right = "";
                for ($j = $i + 1; $j < $len; $j++) {
                    $right .= $splitResult[$j]["text"];
                }

                if (!$this->_isInLiteral($left, $right)) {
                    array_push($tmp, $item);
                } else {
                    array_push($tmp, array("text" => $item["text"], "sep" => null));
                }
            } else {
                array_push($tmp, $item);
            }
        }

        $buffer = "";

        for ($i = 0; $i < $len; $i++) {

            if ($tmp[$i]["sep"] == null) {
                $buffer .= $tmp[$i]["text"];
            } else {
                if (strlen($buffer) > 0) {
                    array_push($res, array("text" => $buffer, "sep" => null));
                    $buffer = "";
                }
                array_push($res, $tmp[$i]);
            }
        }

        if (strlen($buffer) > 0) {
            array_push($res, array("text" => $buffer, "sep" => null));
        }

        return $res;
    }

    private function _isInLiteral($left, $right) {

        foreach ($this->_literalDelims as $delim) {

            $nleft = $this->_numDelimOccurrences($delim, $left);
            if ($nleft % 2 == 0)
                continue;

            $nright = $this->_numDelimOccurrences($delim, $right);
            if ($nright % 2 != 0) {
                return TRUE;
            }
        }

        return FALSE;
    }

    private function _numDelimOccurrences($delim, $text) {

        $res = 0;
        $len = strlen($text);
        $prev = null;

        for ($i = 0; $i < $len; $i++) {
            $ch = $text[$i];
            if ($ch == $delim && ($prev == null || $prev != $this->_literalEscChar)) {
                $res++;
            }
            $prev = $ch;
        }

        return $res;
    }

    private $_literal;
    private $_literalDelims;
    private $_literalEscChar;
    private $_separators;

}

?>

File: src/runtime/python/tokenizer.py
#! coding=UTF-8

# Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#   http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

class Tokenizer(object):

    def __init__(self):

        self._literal = None
        self._literalDelims = ['"', "'"]
        self._literalEscChar = '\\'
        self._separators = []

    def set_literal(self, literal):

        self._literal = literal
        self._literalDelims = literal.DELIMITERS
        self._literalEscChar = literal.ESCAPE_CHAR

    def add_separator(self, sep):

        self._separators.append(sep)

    def split_at_separators(self, text):

        worklist = [(text, None)]

        for sep in self._separators:

            tmp = []
            for text_, sep_ in worklist:
                if sep_:
                    tmp.append((text_, sep_))
                else:
                    tmp += self._split_at_sep(text_, sep)

            worklist = tmp

        return worklist

    def _split_at_sep(self, text, sep):

        regex = sep.getRegexIgnoreWS()

        sep_positions = []
        for m in regex.finditer(text):
            sep_positions.append((m.start(),m.end()))

        tmp = []
        current = 0
        for start, end in sep_positions:
            substr = text[current:start]
            if substr:
                tmp.append((substr, None)) # None --> not a separator
            tmp.append((text[start:end], sep))
            current = end

        substr = text[current:]
        if substr:
            tmp.append((substr, None))

        res = self._unsplit_if_in_literal(tmp)

        return res

    def _unsplit_if_in_literal(self, split_result):

        if not self._literal:
            return split_result

        tmp = []
        last_idx = len(split_result) - 1

        for idx, item in enumerate(split_result):

            substr, sep = item

            if sep is None:
                tmp.append((substr, sep))
                continue
            
            if idx > 0 and idx < last_idx:
                left = "".join([el[0] for el in split_result[:idx]])
                right = "".join([el[0] for el in split_result[idx+1:]])
                if not self._is_in_literal(left, right):
                    tmp.append((substr, sep))
                else:
                    tmp.append((substr, None))
            else:
                tmp.append((substr, sep))

        res = []
        buffer = ""
        for substr, sep in tmp:
            if sep is None:
                buffer += substr
            else:
                if buffer:
                    res.append((buffer, None))
                    buffer = ""
                res.append((substr, sep))
        if buffer:
            res.append((buffer, None))

        return res

    def _is_in_literal(self, left, right):

        for delim in self._literalDelims:
            nleft = self._num_delim_occurrences(delim, left)
            if nleft % 2 == 0:
                continue
            nright = self._num_delim_occurrences(delim, right)
            if nright % 2 != 0:
                return True

        return False

    def _num_delim_occurrences(self, delim, text):

        res = 0
        prev = ''
        for ch in text:
            if ch == delim and (not prev or prev != self._literalEscChar):
                res += 1
            prev = ch

        return res
